src/java.base/share/classes/java/lang/foreign/FunctionDescriptor.java

```diff
@@ -69,7 +70,7 @@ private FunctionDescriptor(MemoryLayout resLayout, List<MemoryLayout> argLayouts
      * {@return the argument layouts associated with this function descriptor}.
      */
     public List<MemoryLayout> argumentLayouts() {
-        return argLayouts;
+        return Collections.unmodifiableList(argLayouts);
```

Comment:
This change doesn’t seem to be necessary, as `FunctionDescriptor` is already created with `List.of(…)` (or `Stream.toList()` in the case of `FunctionDescriptor.VariadicFunction`), which produce immutable lists (although `Stream.toList()` permits `null`s, which `Stream.collect(Collectors.toImmutableList())` and `List.of(…)` doesn’t).

Reply:
> This change doesn’t seem to be necessary, as `FunctionDescriptor` is already created with `List.of(…)` (or `Stream.toList()` in the case of `FunctionDescriptor.VariadicFunction`), which produce immutable lists (although `Stream.toList()` permits `null`s, which `Stream.collect(Collectors.toImmutableList())` and `List.of(…)` doesn’t).

I believe `FunctionDescriptor::insertArgumentLayout` is problematic.

Reply:
Drat, missed that. It's tempting to modify `insertArgumentLayout` and take the copy hit. In general we are lacking good list concatenation functionality.

Reply:
For `insertArgumentLayout`, we could do:
```java
SharedSecrets.getJavaUtilCollectionAccess().listFromTrustedArray(newLayouts.toArray());
```